hack/generator/pkg/astmodel/storage_conversion_factories.go

```diff
@@ -913,21 +850,52 @@ func createTypeDeclaration(name TypeName, generationContext *CodeGenerationConte
 	return astbuilder.Selector(dst.NewIdent(packageName), name.Name())
 }
 
-func zeroValue(p *PrimitiveType) string {
-	switch p {
-	case StringType:
-		return "\"\""
-	case IntType:
-		return "0"
-	case FloatType:
-		return "0"
-	case UInt32Type:
-		return "0"
-	case UInt64Type:
-		return "0"
-	case BoolType:
-		return "false"
-	default:
-		panic(fmt.Sprintf("unexpected primitive type %q", p.String()))
+func zeroValue(t Type, types Types) string {
+	if isTypeOptional(t) {
+		return "nil"
 	}
+
+	if name, isName := AsTypeName(t); isName {
+		// We've got a type name, need to see what it resolves to
+		actualType, err := types.FullyResolve(name)
+		if err != nil {
+			// This should never happen
+			panic(err)
+		}
+
+		enumType, isEnum := AsEnumType(actualType)
+		if isEnum {
+			// Zero value for an enum is the zero value of the base type
+			return zeroValue(enumType.baseType, types)
+		}
+
+		primitiveType, isPrimitive := AsPrimitiveType(actualType)
+		if isPrimitive {
+			return zeroValue(primitiveType, types)
+		}
+
+		// Otherwise default to an empty instantiation
+		return fmt.Sprintf("%s{}", name.Name())
+	}
+
+	if primitiveType, isPrimitive := AsPrimitiveType(t); isPrimitive {
+		switch primitiveType {
+		case StringType:
+			return "\"\""
+		case IntType:
+			return "0"
+		case FloatType:
+			return "0"
+		case UInt32Type:
+			return "0"
+		case UInt64Type:
+			return "0"
+		case BoolType:
+			return "false"
+		default:
+			panic(fmt.Sprintf("unexpected primitive type %q", t.String()))
+		}
```

Comment:
It would be clearer to extract this to `primitiveZeroValue` (and replace the recursive call on 874 with a call to it as well).

Reply:
I simplified this in a different way.